klib/userdata.c

```diff
@@ -0,0 +1,707 @@
+        .userdata_header_name = ss_static_init("Metadata-Flavor"),
+        .userdata_header_value = ss_static_init("Google"),
+        .needs_base64_decode = false,
+        .needs_token = false,
+    },
+    [PROVIDER_AZURE] = {
+        .name = ss_static_init("Azure"),
+        .detect_path = ss_static_init("/metadata/instance?api-version=2021-01-01"),
+        .detect_header_name = ss_static_init("Metadata"),
+        .detect_header_value = ss_static_init("true"),
+        .detect_method = HTTP_REQUEST_METHOD_GET,
+        .userdata_path = ss_static_init("/metadata/instance/compute/userData?api-version=2021-01-01&format=text"),
+        .userdata_header_name = ss_static_init("Metadata"),
+        .userdata_header_value = ss_static_init("true"),
+        .needs_base64_decode = true,
+        .needs_token = false,
+    },
+    [PROVIDER_DIGITALOCEAN] = {
+        .name = ss_static_init("DigitalOcean"),
+        .detect_path = ss_static_init("/metadata/v1/id"),
+        .detect_header_name = sstring_null_init,
+        .detect_header_value = sstring_null_init,
+        .detect_method = HTTP_REQUEST_METHOD_GET,
+        .userdata_path = ss_static_init("/metadata/v1/user-data"),
+        .userdata_header_name = sstring_null_init,
+        .userdata_header_value = sstring_null_init,
+        .needs_base64_decode = true,
+        .needs_token = false,
+    },
+    [PROVIDER_OPENSTACK] = {
+        .name = ss_static_init("OpenStack"),
+        .detect_path = ss_static_init("/openstack/latest/"),
+        .detect_header_name = sstring_null_init,
+        .detect_header_value = sstring_null_init,
+        .detect_method = HTTP_REQUEST_METHOD_GET,
+        .userdata_path = ss_static_init("/openstack/latest/user-data"),
+        .userdata_header_name = sstring_null_init,
+        .userdata_header_value = sstring_null_init,
+        .needs_base64_decode = false,
+        .needs_token = false,
+    },
+    [PROVIDER_HETZNER] = {
+        .name = ss_static_init("Hetzner"),
+        .detect_path = ss_static_init("/hetzner/v1/metadata"),
+        .detect_header_name = sstring_null_init,
+        .detect_header_value = sstring_null_init,
+        .detect_method = HTTP_REQUEST_METHOD_GET,
+        .userdata_path = ss_static_init("/hetzner/v1/userdata"),
+        .userdata_header_name = sstring_null_init,
+        .userdata_header_value = sstring_null_init,
+        .needs_base64_decode = false,
+        .needs_token = false,
+    },
+};
+
+typedef struct userdata {
+    status_handler complete;
+    enum cloud_provider current_provider;
+    buffer aws_token;  /* for AWS IMDSv2 */
+    boolean detecting;
+    boolean started;  /* true after first provider attempt */
+    int retry_count;
+    buffer save_file_path;  /* path to save raw userdata if configured */
+    closure_struct(value_handler, userdata_vh);
+    closure_struct(value_handler, detect_vh);
+} *userdata;
+
+static userdata ue;
+static heap userdata_heap;
```

Comment:
This heap pointer can be embedded in the userdata structure, no need for a separate variable